### DISCUSSION TITLE
Wifi on Pico W board doesn't work when the system frequency is too high, the program remains in an infinite loop

### DISCUSSION THREAD

**Discussion by klug27** [earlephilhower/arduino-pico#2590] [Q&A] [answered]
Thank you for all person who have participated to build these powerfull repository. Please i have some problem with the pico w board. When i set the system frequency more than 250Mhz the program loop. To correct these i have increase the prescaler value of the pio spi wifi program to 4. This program no longer loops permanently with the pico example program, but with Arduino the program always loops.
![Arduino_main](https://github.com/user-attachments/assets/f0dd454e-6961-4c41-a3fa-7ad8e5a29a01)
![Arduino_wifi](https://github.com/user-attachments/assets/9b0a12be-6d3a-4309-beee-1c5cfa641fe0)
![pico_sdk](https://github.com/user-attachments/assets/6b9ae0b5-b319-4f18-b47c-68618fe73c51)
![pico_sdk_main](https://github.com/user-attachments/assets/8074f621-a3b8-43b5-bac1-ad9411129724)

 Please how can i fix this problem ?

**Comment by maxgerhardt** [comment]
If you need to apply a fix in `cyw43_bus_pio_spi.c` in the SDK files, then for Arduino-Pico, you would have to recompile the Pico-SDK library file; Pico-SDK is not built from source, but taken from the `.a` file instead. https://github.com/earlephilhower/arduino-pico/tree/master/tools/libpico has instructions on how to do that. Note, that you would just need to define `CYW43_PIO_CLOCK_DIV=4` ([source](https://github.com/raspberrypi/pico-sdk/blob/efe2103f9b28458a1615ff096054479743ade236/src/rp2_common/pico_cyw43_driver/include/pico/cyw43_driver.h#L48-L55)) in the CMake file, so no C file changes should be necessary, only compile config.

Alternatively, CC @earlephilhower, we could activate the `CYW43_PIO_CLOCK_DIV_DYNAMIC` (=1) option in the SDK to make the `cyw43_set_pio_clock_divisor()` function available, which the user could then call (or the core) in case of an overclocked CPU, to still get correct WiFi SPI communication.

https://github.com/raspberrypi/pico-sdk/blob/efe2103f9b28458a1615ff096054479743ade236/src/rp2_common/pico_cyw43_driver/include/pico/cyw43_driver.h#L42-L64

We would have to add that probably in https://github.com/earlephilhower/arduino-pico/blob/master/tools/libpico/CMakeLists.txt#L49-L59.

**Reply by maxgerhardt** [reply]
Related to https://github.com/earlephilhower/arduino-pico/pull/2596 now

**Reply by maxgerhardt** [reply, marked as answer]
I've managed to successfully recompile libpico with `CYW43_PIO_CLOCK_DIV_DYNAMIC=1` and add `cyw43_set_pio_clock_divisor(4, 0); // div by 4.0` into `picow_init.cpp` before the `cyw43_arch_init_with_country()`, and my example WiFi scan sketch with a Pico clocked at 260MHz went from crashing after printing the clock speed to working successfully and scanning the WiFi networks.

I could not test 275 MHz or 300 MHz because at that speed, the USB connection crashed / was not established.

Example PlatformIO project: https://github.com/maxgerhardt/pio-overclocked-picow-test
The arduino-pico fork: https://github.com/maxgerhardt/arduino-pico/tree/overclocked_wifi

**Reply by klug27** [reply]
Thank you very much

**Reply by maxgerhardt** [reply]
FYI the fix for div by 4 has been added into mainline some time ago, please don't use the above Arduino-Pico fork or project. 

https://github.com/earlephilhower/arduino-pico/blob/6fc3165d3a6024d221ed8897cf2b6f52c613fb10/cores/rp2040/cyw43_wrappers.cpp#L149-L159

If you just use the latest version of the Arduino-Pico core, it will already be fixed.
